### PR TITLE
[ML] Adding OpenAI service to inference API

### PR DESCRIPTION
**Pull request by jonathan-buttner**
WIP

This PR adds functionality to interact with openai for text embedding.

OpenAI text embedding API docs: https://platform.openai.com/docs/api-reference/embeddings/create

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpClient.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.InferencePlugin.MAX_HTTP_RESPONSE_SIZE;
+
+public class HttpClient {
+    // TODO pick a reasonable value for this
+    private static final int MAX_CONNECTIONS = 500;
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpClient.java:34]: `MAX_CONNECTIONS` could be a setting

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/SizeLimitInputStream.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        byteCounter.incrementAndGet();
+        checkMaximumLengthReached();
+        return super.read();
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        byteCounter.addAndGet(len);
+        checkMaximumLengthReached();
+        return super.read(b, off, len);
```
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/SizeLimitInputStream.java:48]: 👍

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferencePlugin.java [hunk 3]
```diff
@@ -155,4 +181,18 @@ public String getFeatureName() {
     public String getFeatureDescription() {
         return "Inference plugin for managing inference services and inference";
     }
+
+    @Override
+    public List<ExecutorBuilder<?>> getExecutorBuilders(Settings unused) {
+        ScalingExecutorBuilder utility = new ScalingExecutorBuilder(
+            UTILITY_THREAD_POOL_NAME,
+            1,
+            512 * 4,
```
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/InferencePlugin.java:190]: > What's the logic behind 512 * 4?

I just blindly copied what we had in [MachineLearning.java](https://github.com/elastic/elasticsearch/blob/bd136f86edd80d24d340b407a35916d74fe65b53/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java#L1632-L1639)

Any thoughts on a reasonable max?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpClient.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    public void send(HttpUriRequest request, ActionListener<HttpResponse> listener) throws IOException {
+        if (client.isRunning() == false) {
+            client.start();
+        }
+
+        SocketAccess.doPrivileged(() -> client.execute(request, new FutureCallback<>() {
+            @Override
+            public void completed(org.apache.http.HttpResponse result) {
+                try {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpClient.java:66]: We likely need to kick off a new thread here

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiEmbeddingsResponseEntity.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+// TODO figure out the format for the response
+public record OpenAiEmbeddingsResponseEntity(Embedding[] embeddings) implements InferenceResult {
+    private static final String NAME = "openai_embeddings_response_entity";
+    private static final String EMBEDDINGS_FIELD = "embeddings";
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiEmbeddingsResponseEntity.java:31]: TODO: Change this to `text_embedding` and confirm with the clients team if this makes sense.

**Comment by jonathan-buttner** [comment]
Closing in favor of: https://github.com/elastic/elasticsearch/pull/101893

**Pull request closed without merging**